src/edges.rs

```diff
@@ -0,0 +1,131 @@
+use image::{Pixel, GenericImage, ImageBuffer, Luma};
+use image::imageops::blur;
+use gradients::{vertical_sobel, horizontal_sobel};
+
+/// Runs the canny edge detection algorithm on the provided `ImageBuffer`.
+///
+/// Returns a binary image, where edge pixels have a value of 255 and non-edge pixels a value of 0.
+pub fn canny<I>(image: &I,
+                low_threshold: u8,
+                high_threshold: u8)
+                -> ImageBuffer<Luma<u8>, Vec<<Luma<u8> as Pixel>::Subpixel>>
+    where I: GenericImage<Pixel = Luma<u8>> + 'static
+{
+    if high_threshold < low_threshold {
+        panic!("Low threshold is greater than high threshold: {}; {}",
+               low_threshold,
+               high_threshold)
+    }
+    // Heavily based on the implementation proposed by wikipedia.
+    // 1. Gaussian blur.
+    const SIGMA: f32 = 1.4;
+    let blurred = blur(image, SIGMA);
+
+    // 2. Intensity of gradients.
+    let gx = horizontal_sobel(&blurred);
+    let gy = vertical_sobel(&blurred);
+    let mut g = ImageBuffer::from_fn(image.width(), image.height(), |x, y| {
+        let g = (gx[(x, y)][0] as f32).hypot(gy[(x, y)][0] as f32);
+        Luma { data: [g as u8] }
+    });
+
```

Comment:
Oh yeah, I see the problem now. Would it be a possible workaround to use u8::MAX if g > 255?

The values are all positive, since we use the hypot function.